src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.ChangeNamespaceCodeAction.cs

```diff
@@ -0,0 +1,482 @@
+                // just ones that fully matche the old or new namespace.
+                // For example, if we are changing namespace `Foo.Bar` to `A.B`, the using of name `Bar` below would remain 
+                // untouched:
+                //
+                //      namespace Foo
+                //      {
+                //          using Bar;
+                //          ~~~~~~~~~
+                //      }
+                //
+
+                var solutionAfterFirstMerge = await MergeDiffAsync(_state.Solution, solutionAfterNamespaceChange, cancellationToken).ConfigureAwait(false);
+
+                var solutionAfterImportsRemoved = await RemoveUnnecessaryImportsAsync(solutionAfterFirstMerge, _state.DocumentIds, CreateAllContainingNamespaces(OldNamespaceParts), cancellationToken)
+                    .ConfigureAwait(false);
+
+                solutionAfterImportsRemoved = await RemoveUnnecessaryImportsAsync(
+                    solutionAfterImportsRemoved, referenceDocuments.ToImmutableArray(), ImmutableArray.Create(_state.DeclaredNamespace, _state.TargetNamespace), cancellationToken)
+                    .ConfigureAwait(false);
+                referenceDocuments.Free();
+
+                var solutionAfterSecondMerge = await MergeDiffAsync(solutionAfterFirstMerge, solutionAfterImportsRemoved, cancellationToken).ConfigureAwait(false);
+
+                return ImmutableArray.Create<CodeActionOperation>(new ApplyChangesOperation(solutionAfterSecondMerge));
+            }
+
+            private ImmutableArray<string> CreateAllContainingNamespaces(ImmutableArray<string> parts)
+            {
+                var builder = ArrayBuilder<string>.GetInstance();
+                for (var i = 1; i <= OldNamespaceParts.Length; ++i)
+                {
+                    builder.Add(string.Join(".", OldNamespaceParts.Take(i)));
+                }
+                return builder.ToImmutableAndFree();
+            }
+
+            private ImmutableArray<SyntaxNode> CreateImports(Document document, ImmutableArray<string> names)
+            {
+                var generator = SyntaxGenerator.GetGenerator(document);
+                var builder = ArrayBuilder<SyntaxNode>.GetInstance(names.Length);
+                for (var i = 0; i < names.Length; ++i)
+                {
+                    builder.Add(generator.NamespaceImportDeclaration(names[i]));
+                }
+                return builder.ToImmutableAndFree();
+            }
+
+            private SyntaxNode CreateImport(Document document, string name)
+            {
+                var syntaxGenerator = SyntaxGenerator.GetGenerator(document);
+                return syntaxGenerator.NamespaceImportDeclaration(name);
+            }
+
+            private async Task<(Solution, ImmutableArray<DocumentId>)> ChangeNamespaceToMatchFoldersAsync(Solution solution, DocumentId id, CancellationToken cancellationToken)
```

Comment:
will review all this later. the compelxity level was very suprising.

Reply:
should be documentd as to what it does.  for example, what are the document-ids being returned.  In general, if hte return values of something are not obvious given the name of hte method, docs are good.